Add blog post for mypy imports (#74)
octoenergy/octoenergy.github.io: _posts/2024-08-19-mypy-imports.markdown

Let's fake the installation of a third party package by navigating to our virtual environment's `site-packages` folder
and creating a package in there named `thirdparty`:

```
$ mkdir .venv/lib/python3.12/site-packages/thirdparty
$ touch .venv/lib/python3.12/site-packages/thirdparty/__init__.py
```

Then, in the file we just created, add that same code with the typing mistake:

```python
def get_string() -> str:
    return 1
```

Next, replace the entire contents of `blue.py` with this code:

```python
import thirdparty

some_string = thirdparty.get_string()
```

Let's see what happens when we type check `blue.py`:

```
$ mypy -m blue
blue.py:1: error: Skipping analyzing "thirdparty": module
is installed, but missing library stubs or py.typed marker
[import-untyped]
blue.py:1: note:
See https://mypy.readthedocs.io/en/stable/running_mypy.html#missing-imports
Found 1 error in 1 file (checked 1 source file)
```

Because mypy thinks of this module as a third party package (i.e. installed in your virtualenv's `site-packages`
rather than alongside `blue.py`), we have a new error. This is because mypy requires the package to declare

> that it is 'fit for type checking' before it will use it. The simplest way to do this is to add an empty file called

Reviewer: 🐼 "before it will use it" is a little ambiguous since we're saying "it" twice in reference to two separate things. May we elaborate what "it" is here?

Author: Good idea, have edited.

`py.typed`, which tells mypy to look at the annotations in that package:

```
$ touch .venv/lib/python3.12/site-packages/thirdparty/py.typed
```

With that in place, let's type check `blue` again:

```
$ mypy -m blue
Success: no issues found in 1 source file
```

So this is interesting. There is an error in the third party library, but mypy is not complaining about it. So if it's

> not type checking our third party package, why is it complaining about that `py.typed` file in the first place?

Author: Good idea, I've updated this.

```python
import copy
import thirdparty
import green

copy.copy()
some_string: int = thirdparty.get_string()
```

And now let's run it:

```
$ mypy -m blue
red.py:2: error: Incompatible return value type (got "int",
expected "str")  [return-value]
blue.py:5: error: Missing positional argument "x" in call to
"copy"  [call-arg]
blue.py:6: error: Incompatible types in assignment (expression
has type "str", variable has type "int")  [assignment]
Found 3 errors in 2 files (checked 1 source file)
```

So, we have three different errors here:

1. An error internal to an upstream local module `red.py` (its `get_string` function is annotated to return a string,
   but actually returns an integer).
2. An error calling a standard library module (`copy` expects an argument).
3. An error calling a third party module (we've annotated a value as an `int`, but the third party library says its
   a `str`).

Although we haven't passed an argument, we're seeing mypy run in `normal` mode: the most 'dogged' mode.
As we've already seen, even in this mode we won't see any errors relating to the internals of third party or standard library packages.

> Let's run it in the mode at the opposite end of the spectrum: `error`:

Reviewer: Would it be worthwhile giving each flag mode in its own sub-heading? It may be easier to follow through and jump back to where a reader leaves off if they're interrupted mid-read.

Author: Great idea, done.

```
$ mypy -m blue --follow-imports=error
blue.py:2: error: Import of "thirdparty" ignored  [misc]
blue.py:2: note: (Using --follow-imports=error, module
not passed on command line)
blue.py:3: error: Import of "green" ignored  [misc]
blue.py:5: error: Missing positional argument "x" in
call to "copy"  [call-arg]
Found 3 errors in 1 file (checked 1 source file)
```

We still have three errors, but this time two of them are simply triggered by the imports themselves.
Notice that mypy is treating the standard library package (`copy`) differently to the third party package &mdash; it's checking
we're interacting correctly with it even though we're passing `error`.

This mode is designed to force callers to explicitly specify all packages for checking, rather than let the imports
be followed implicitly. To get around it, we can tell mypy to check all of the modules by passing all of them like this:

```python

> mypy -m blue -m green -m red -m thirdparty --follow-imports=error

Reviewer 2: Maybe don't mark the fenced code block as `python` since it is shell?
Add `$` before the command to indicate that this should be entered (and also for consistency with the similar fenced code blocks in the article)?

Author: Oops, yes definitely. Thanks.

because we're skipping it, mypy is not bothering even to check the type signature of that function. We'd see the same
behaviour if we were interacting incorrectly with a local module: no error!

Finally, there is `silent` mode:

```
mypy -m blue --follow-imports=silent
blue.py:5: error: Missing positional argument "x" in call
to "copy"  [call-arg]
blue.py:6: error: Incompatible types in assignment (expression
has type "str", variable has type "int")  [assignment]
Found 2 errors in 1 file (checked 1 source file)
```

This mode sits between `normal` and `skip`. What's happening under the hood is that mypy is following the imports, but
it's not reporting on errors outside the modules we explicitly ask to check. This has the effect of treating local
modules (that weren't passed) in the same way as the standard library and third party packages are: internal errors
aren't reported on, but mistakes relating to interacting with those modules *are* reported as errors.

## Observations on the different modes

[The mypy docs](https://mypy.readthedocs.io/en/stable/running_mypy.html#following-imports) recommend using `normal`

> or `error` modes, if possible, as it makes sure you're accidentally skipping any part of the code base. The interesting

Author: 🤦🏻

> thing about `error` mode is it will force you to check the internals of all third party packages, too, which seems

Reviewer: Can we think of a case where `error` would be valuable, even if we think for most users it's unlikely to be useful?

Author: I mean, I can't really, that's my point. I think it's probably okay for the post to be subjective, but if anyone can think of an idea then let me know.

I actually wonder if the _Observations_ section would be better off deleted as it's perhaps not that interesting, and makes the piece unnecessarily subjective.

strange to me.

`skip`, the docs suggests, should be used with caution. I can understand why: detecting
whether a function in an upstream package is called with the correct types is, surely, essential if you're going to bother
running a type checker at all? But perhaps there is some niche use case I haven't thought of.

> Finally, there's `silent`. The mypy docs suggest this is a kind of compromise mode, to be used if it is just too

Reviewer: 🐼 may be able to shorten this sentence for easier reading to something like below:

```suggestion
Finally, there's `silent`. The mypy docs suggest this is a compromise mode to use if it is just too
```

Author: Thanks, have edited.

to use a type stub over a real module.

The other interesting thing is that, despite being in `follow-imports=normal` mode, there is no error from `red.py`.
This is, again, because mypy isn't looking at `green.py` at all, only at the stub file. If we did import `red` from
the stub file then the error would show up again.

There are more rules around the discovery of stub files that I won't go into here. The key take away is that stub files
allow us to provide alternatives for what mypy looks at &mdash; and, therefore, which modules it views as upstream, when
following imports.

## In summary

When type checking a module, mypy will often encounter an import of a module that it hasn't been explicitly asked to
type check. It does different things depending on the kind of module, the `follow-imports` mode we tell it to run in,
and whether it finds a stub file or not.

- Mypy _always_ checks that we are interacting correctly with the standard library
 (and _never_ checks standard library internals).
- In `normal` mode, third party packages are checked in the same way as standard library packages.

> - Local packages are different: in `normal` mode, mypy follows the chain of imports and type checks their internals too.

Reviewer: 🐼 should this point perhaps be nested in with the above point so it's clear both relate to `normal` mode?
```suggestion
    - Local packages are different: in `normal` mode, mypy follows the chain of imports and type checks their internals too.
```

Author: Good idea.

- `silent` mode makes unpassed local packages behave like standard library / third party packages: it won't error
  on their internals, only on the interactions.
- `skip` mode doesn't check interactions with anything except the standard library.

> - `error` mode errors if we try to import anything (except the standard library) that isn't explicitly passed.

Reviewer 3: Would be curious to know of the performance impact for these on our code base, I presume `skip` or `error` is the fastest and `normal` the slowest but by how much

Author: I'm going to leave this out of this blog post as I think it is one thing too many...
